### PR TITLE
Tgp 2005/store records changes

### PR DESCRIPTION
**Pull request by besscerule**

### CODE REVIEW

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/controllers/actions/StoreLatestAction.scala [hunk 1]
```diff
@@ -52,7 +52,13 @@ class StoreLatestActionImpl @Inject() (
             case None         => None
           }
         )
-        .map(_ => None) transform {
+        .map(isDone =>
```
[the remark this answers is not in the document]
Comment (wolfendale) [reply app/uk/gov/hmrc/tradergoodsprofilesdatastore/controllers/actions/StoreLatestAction.scala:55]:
```scala
response match {
  case Some(record) => Some(record.updatedDateTime.toString)
  case None                => None
}
```
Can be simplified to
```scala
response.map(_.updatedDateTime.toString)
```
Although as above this should be coming from the `RecordSummary` instead

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala [hunk 2]
```diff
@@ -18,202 +18,178 @@ package uk.gov.hmrc.tradergoodsprofilesdatastore.services
...
           )
         )
-
-        val result = await(service.storeRecords(requestEori, None)(FakeRequest(), hc))
-
-        result.value shouldBe Done
-        verify(mockRouterConnector, times(7)).getRecords(any(), any(), any(), any())(any())
-        verify(mockRecordsRepository, times(6)).saveRecords(any(), any())
-        verify(mockRecordsRepository).getCountWithInactive(any())
+        val result = await(service.storeRecords(requestEori, None)(hc))
```
Comment (wolfendale) [inline test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala:101]: Should be able to do
```scala
result shouldBe false
```
here instead of in the `eventually` as this indicates that it's separate from the things that are happening concurrently

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala [hunk 3]
```diff
@@ -19,57 +19,92 @@ package uk.gov.hmrc.tradergoodsprofilesdatastore.services
...
-                s"Data Store and B&T Database are out of sync: There are $totalDataStoreRecords data store records and $totalRouterRecords B&T records."
-              )
-            )
-          }
+  )(implicit hc: HeaderCarrier): Future[Boolean] =
+    storeFirstBatchOfRecords(eori, lastUpdatedDate).flatMap { recordsToStore =>
+      if (recordsToStore > recursivePageSize) {
+        recordsSummaryRepository.set(eori, Some(Update(recursivePageSize, recordsToStore - recursivePageSize))).map {
+          _ =>
+            storeRecordsRecursively(eori, recursiveStartingPage + 1, lastUpdatedDate, 0, recordsToStore).onComplete {
```
Comment (AmarkumarJagtap) [inline app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala:46]: does the recordsToStore should be before 0, as 0 is recordsStored ?

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala [hunk 4]
```diff
@@ -19,57 +19,92 @@ package uk.gov.hmrc.tradergoodsprofilesdatastore.services
...
+    recordsToStore: Int,
+    recordsStored: Int
+  )(implicit hc: HeaderCarrier): Future[Done] =
     routerConnector.getRecords(eori, lastUpdatedDate, Some(page), Some(recursivePageSize)).flatMap { recordsResponse =>
       recordsRepository.saveRecords(eori, recordsResponse.goodsItemRecords).flatMap { _ =>
-        if (recordsResponse.pagination.nextPage.isDefined) {
-          storeRecordsRecursively(eori, page + 1, lastUpdatedDate)
-        } else {
-          Future.successful(Done)
+        val newRecordsToStore = recordsToStore - recordsResponse.goodsItemRecords.size
```
Comment (AmarkumarJagtap) [inline app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala:100]: its coming here negative because of above

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala
Comment (wolfendale) [inline test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala:1]: We're missing a test to show that the update is set to `None` even when the recursive call fails.

These tests are pretty tough in unit tests, it may make more sense to test this with actual mongo collections so that we can test the state of things

**Review by wolfendale — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala (hunk 4 above)
Comment (besscerule) [reply app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala:105]: fixed in both places please check

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala (hunk 3 above)
Comment (besscerule) [reply app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala]: fixed in both places please check

**Review by AmarkumarJagtap — APPROVED** [review]
LGTM

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala [hunk 5]
```diff
@@ -19,57 +19,92 @@ package uk.gov.hmrc.tradergoodsprofilesdatastore.services
...
+                }
+            }
+            false
+        }
+      } else {
+        checkInSync(eori).map(_ => true)
+      }
+    }
+
+  private def checkInSync(eori: String)(implicit hc: HeaderCarrier): Future[Done] =
```
[the remark this answers is not in the document]
Comment (besscerule) [reply app/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsService.scala:64]: with only storing active records this becomes a bit of a problem anyway.. I say we leave it for the moment and then when we look at the inactive delete (see next PR) we can review that then

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala (hunk 2 above)
Comment (besscerule) [reply test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala:101]: if i don't do it in the eventually it doesn't complete and errors - i think i need a success or assertion at the end of the eventually block to make it complete, this is why i put the result should be false there

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala
Comment (besscerule) [reply test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala:1]: please can you help with this bit!

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/actions/StoreLatestActionSpec.scala
[the remark this answers is not in the document]
Comment (besscerule) [reply test/uk/gov/hmrc/tradergoodsprofilesdatastore/actions/StoreLatestActionSpec.scala:1]: done

File: app/uk/gov/hmrc/tradergoodsprofilesdatastore/controllers/actions/StoreLatestAction.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (besscerule) [reply app/uk/gov/hmrc/tradergoodsprofilesdatastore/controllers/actions/StoreLatestAction.scala:51]: done

File: test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala
Comment (AmarkumarJagtap) [reply test/uk/gov/hmrc/tradergoodsprofilesdatastore/services/StoreRecordsServiceSpec.scala:1]: added test case to check oncomplete that sets update to None

**Comment by wolfendale** [comment]
Closing this as all changes are in #39
